src/bridge/cockpitbridge.c

```diff
@@ -0,0 +1,271 @@
+
+      channel_function = g_hash_table_lookup (self->payloads, payload);
+      if (channel_function != NULL)
+        {
+          channel_type = channel_function ();
+        }
+
+      g_debug ("Running channel type %lu for %s default is %lu", channel_type, payload, COCKPIT_TYPE_CHANNEL);
+      channel = g_object_new (channel_type,
+                              "transport", transport,
+                              "id", channel_id,
+                              "options", options,
+                              NULL);
+      g_hash_table_insert (self->channels, g_strdup (channel_id), channel);
+      g_signal_connect (channel, "closed", G_CALLBACK (on_channel_closed), self);
+    }
+}
+
+static gboolean
+on_transport_control (CockpitTransport *transport,
+                      const char *command,
+                      const gchar *channel_id,
+                      JsonObject *options,
+                      GBytes *message,
+                      gpointer user_data)
+{
+  CockpitBridge *self = user_data;
+
+  if (g_str_equal (command, "init"))
+    {
+      process_init (self, transport, options);
+      return TRUE;
+    }
+
+  if (!self->init_received)
+    {
+      g_warning ("caller did not send 'init' message first");
+      cockpit_transport_close (transport, "protocol-error");
+      return TRUE;
+    }
+
+  if (g_str_equal (command, "open"))
+    {
+      process_open (self, transport, channel_id, options);
+      return TRUE;
+    }
+  else if (g_str_equal (command, "close"))
+    {
+      if (!channel_id)
+        {
+          g_warning ("Caller tried to close channel without an id");
+          cockpit_transport_close (transport, "protocol-error");
+        }
+      else
+        {
+          /*
+           * The channel may no longer exist due to a race of the bridge closing
+           * a channel and the web closing it at the same time.
+           */
+
+          g_debug ("already closed channel %s", channel_id);
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+cockpit_bridge_add_payload (CockpitBridge *self,
+                            const gchar *type,
+                            type_function channel_function)
+{
+  return g_hash_table_insert (self->payloads, g_strdup (type),
+                              channel_function);
+}
+
+CockpitBridge *
+cockpit_bridge_new (CockpitTransport *transport,
+                    CockpitPayloadType *payload_types,
+                    gboolean init_received)
+{
+  gint i;
+  CockpitBridge *bridge  = g_object_new (COCKPIT_TYPE_BRIDGE,
+                                         "init-received", init_received,
+                                         NULL);
+
+  for (i = 0; payload_types[i].name != NULL; i++)
+    {
+      cockpit_bridge_add_payload (bridge, payload_types[i].name,
+                                  payload_types[i].function);
+    }
+  g_signal_connect (transport, "control", G_CALLBACK (on_transport_control), bridge);
```

Comment:
So the bridge should last until the channel is closed? If not, then we should probably comment here how the if the bridge is closed, but the transport is still open, then things will go boom.